File: Linked List/2. Doubly Linked List/3. Doubly LinkedList Deletion/DoublyLinkedListDeletion.cpp
#include <iostream>
using namespace std;

class DLLNode
{
    public:
        int data;
        DLLNode* nextElement;
        DLLNode* prevElement;

        DLLNode()
        {
            nextElement = nullptr;
            prevElement = nullptr;
        }
};

class DoublyLinkedList
{
    private:
        DLLNode* head;
    
    public:
        DoublyLinkedList()
        {
            head = nullptr;
        }

        bool isEmpty()
        {
            if (head == nullptr)
                return true;
            else
            {
                return false;
            }
            
        }

        void printDLL()
        {
            if (isEmpty())
            {
                cout << "Doubly DoublyLinkedList is empty." << endl;
            }

            DLLNode* temp = head;
            while (temp != nullptr)
            {
                cout << temp->data << "->";
                temp = temp->nextElement;
            }
            cout << "null " << endl;
        }

        void insertAtHead(int value)
        {
            DLLNode* newNode = new DLLNode();
            newNode->data = value;

            newNode->nextElement = head;

            if (head != nullptr)
                head->prevElement = newNode;
            
            head = newNode;
            cout << value << " inserted." << endl;

        }

        void insertAtTail(int value)
        {
            DLLNode* newNode = new DLLNode();
            newNode->data = value;

            DLLNode* temp = head;

            while (temp->nextElement != nullptr)
                temp = temp->nextElement;

            temp->nextElement = newNode;
            newNode->prevElement = temp;

            cout << value << " inserted." << endl;
        }

        void deleteANode(int value)
        {
            if (isEmpty())
            {
                cout << "List is empty." << endl;
                return;
            }

            DLLNode* currentNode = head;

            if (currentNode->data == value)
            {
                head = head->nextElement;
                head->prevElement = nullptr;
                cout << value << " deleted." << endl;
                return;
            }

            while (currentNode != nullptr)
            {
                if (currentNode->data == value)
                {
                    currentNode->prevElement->nextElement = currentNode->nextElement;
                    if(currentNode->nextElement != nullptr)
                    {
                        currentNode->nextElement->prevElement = currentNode->prevElement;
                    }

                    cout << value << " deleted." << endl;
                    return;
                }
                currentNode = currentNode->nextElement;
            }

            if (currentNode == nullptr)
                cout << value << " does not exist." << endl;            
        }
};

int main()
{
    DoublyLinkedList list; // DoublyLinkedList created

    for (int i = 0; i < 5; i++)
    {
        list.insertAtHead(i+1);
        list.printDLL();
    }

    for (int i = 30; i > 25; i--)
    {
        list.insertAtTail(i);
        list.printDLL();
    }

    list.deleteANode(3);
    list.printDLL();
    list.deleteANode(43);
    list.printDLL();


    return 0;
}

File: Linked List/2. Doubly Linked List/3. Doubly LinkedList Deletion/DoublyLinkedListDeletion.java
class DLLNode
{
    int value;
    DLLNode next;
    DLLNode prev;

    DLLNode()
    {
        this.next = null;
        this.prev = null;
    }

    DLLNode(int value)
    {
        this.value = value;
        this.next = null;
        this.prev = null;
    }
}

public class DoublyLinkedListDeletion
{
    private DLLNode head;

    public DoublyLinkedListDeletion()
    {
        head = null;
    }

    public boolean isEmpty()
    {
        if (head == null)
            return true;
        else
            return false;
    }

    public void printList()
    {
        if (isEmpty())
            System.out.println("LinkedList is Empty");
        
        DLLNode temp = head;

        while (temp != null)
        {
            System.out.print(temp.value + "->");
            temp = temp.next;
        }
        System.out.println("null");
    }

    public void insertAtHead(int value)
    {
        DLLNode newNode = new DLLNode(value);

        newNode.next = head;

        if (head != null)
            head.prev = newNode;
        
        head = newNode;
        System.out.println(value + " inserted." );
    }

    public void insertAtTail(int value)
    {
        DLLNode newNode = new DLLNode(value);
        
        if (isEmpty())
        {
            head = newNode;
            System.out.println(value + " inserted.");
            return;
        }

        DLLNode temp = head;
        while (temp.next != null)
        {
            temp = temp.next;
        }

        temp.next = newNode;
        newNode.prev = temp;

        System.out.println(value + " inserted.");
    }

    public void deleteANode(int value)
    {
        if (isEmpty())
        {
            System.out.println("List is Empty.");
            return;
        }

        DLLNode temp = head;
        if (temp.value == value)
        {
            head = head.next;
            head.prev = null;
            System.out.println(value + " deleted.");
            return;
        }

        while (temp != null)
        {
            if(temp.value == value)
            {
                temp.prev.next = temp.next;
                if (temp.next != null)
                {
                    temp.next.prev = temp.prev;
                }
                System.out.println(value + " deleted.");
                return;
            }
            temp = temp.next;
        }

        if (temp == null)
        {
            System.out.println(value + " does not exist.");
        }
    }

    public static void main(String args[])
    {
        DoublyLinkedListDeletion list = new DoublyLinkedListDeletion();

        for (int i = 0; i < 8; i++)
        {
            list.insertAtHead(i);
            list.printList();
        }

        for (int i = 30; i > 25; i--)
        {
            list.insertAtTail(i);
            list.printList();
        }

        list.deleteANode(8);
        list.deleteANode(27);
        list.printList();
        list.deleteANode(7);
        list.printList();
    }
}